kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/InternalCompletableFuture.java

```diff
@@ -73,9 +77,242 @@ public CompletionStage<T> minimalCompletionStage() {
      * @param <U> the type of the value
      * @return the completed CompletableFuture
      */
-    public static <U> CompletableFuture<U> completedFuture(Executor executor, U value) {
+    public static <U> CompletableFuture<U> completedFuture(EventLoop executor, U value) {
         var f = new InternalCompletableFuture<U>(executor);
         f.complete(value);
         return f;
     }
+
+    @Override
+    public <U> CompletableFuture<U> thenApply(Function<? super T, ? extends U> fn) {
+        return super.thenCompose(t -> {
+            if (isInEventLoop()) {
+                return super.thenApply(fn);
+            }
+            else {
+                return super.thenApplyAsync(fn);
+            }
+        });
+    }
+
+    private boolean isInEventLoop() {
+        return eventLoop.inEventLoop();
+    }
+
+    @Override
+    public CompletableFuture<Void> thenAccept(Consumer<? super T> action) {
+        return super.thenCompose(t -> {
+            if (isInEventLoop()) {
+                return super.thenAccept(action);
+            }
+            else {
+                return super.thenAcceptAsync(action);
+            }
+        });
+    }
+
+    @Override
+    public CompletableFuture<Void> thenRun(Runnable action) {
+        return super.thenCompose(t -> {
+            if (isInEventLoop()) {
+                return super.thenRun(action);
+            }
+            else {
+                return super.thenRunAsync(action);
+            }
+        });
+    }
+
+    @Override
+    public <U, V> CompletableFuture<V> thenCombine(CompletionStage<? extends U> other, BiFunction<? super T, ? super U, ? extends V> fn) {
+        CompletionStage<? extends U> internalFuture = toInternalFuture(other);
+        return super.thenCombineAsync(internalFuture, fn);
+    }
+
+    private <U> CompletionStage<? extends U> toInternalFuture(CompletionStage<? extends U> other) {
+        if (other instanceof InternalCompletableFuture) {
+            return other;
+        }
+        CompletableFuture<U> incompleteFuture = newIncompleteFuture();
+        other.whenComplete((u, throwable) -> {
+            if (eventLoop.inEventLoop()) {
```

Comment:
not sure if moving the completion to the eventLoop is really necessary here, since the resulting stage is an internal future with all the guarantees about where work will execute.